src/message.rs

```diff
@@ -36,14 +36,160 @@ impl Deserialize for Version {
+
+    () => {
+        None
+    };
+
+    // Review: do we want to use this to send a `"params": null` as opposed to skipping it?
+    //         params!() or None can be the skip while params!(null) will actually send null
+    //         Notes:
+    //           - Params::Null needs to be added
+    //           - `None` and `Some(Params::Null)` are "almost" equivalent on server side
+    (null) => {
+        None
+    };
+
+    (true) => {
+        Some($crate::Params::Positional(vec![$crate::macro_exports::Value::Bool(true)]))
+    };
+
+    (false) => {
+        Some($crate::Params::Positional(vec![$crate::macro_exports::Value::Bool(false)]))
+    };
+
+    ([]) => {
+        {
+            Some($crate::Params::Positional(vec![]))
+        }
+    };
+
+    ([ $($tt:tt)+ ]) => {
+        {
+            match $crate::macro_exports::Value::Array(json_internal!(@array [] $($tt)+)) {
+                $crate::macro_exports::Value::Array(xs) => Some($crate::Params::Positional(xs)),
+                _ => unreachable!(),
+            }
+        }
+    };
+
+    ({}) => {
+        {
+            Some($crate::Params::Named($crate::Map::new()))
+        }
+    };
+
+    ({ $($tt:tt)+ }) => {
+        {
+            let mut object = $crate::macro_exports::Map::new();
+            json_internal!(@object object () $($tt)+);
+            Some($crate::Params::Named(object))
+        }
+    };
+
+    // Force structs to be serialized as positional
+    (pos $other:expr) => {
+        match $crate::macro_exports::to_value(&$other).unwrap() {
+            // Turn the object into positional discarding the names
+            $crate::macro_exports::Value::Object(x) => {
+                let mut xs = Vec::with_capacity(x.len());
+                for (_,v) in x {
+                    xs.push(v);
+                }
+                Some($crate::Params::Positional(xs))
+            },
+            $crate::macro_exports::Value::Array(xs) =>
+                Some($crate::Params::Positional(xs)),
+            // if the object serializes to `Null` map it to `None`
+            $crate::macro_exports::Value::Null => None,
+            // encode any other value as an array of 1 item
+            value => Some($crate::Params::Positional(vec![value])),
+        }
+    };
+
+    // Any Serialize type: numbers, strings, struct literals, variables etc.
+    // Must be below every other rule.
+    ($other:expr) => {
+        match $crate::macro_exports::to_value(&$other).unwrap() {
+            $crate::macro_exports::Value::Object(x) =>
+                Some($crate::Params::Named(x)),
+            $crate::macro_exports::Value::Array(xs) =>
+                Some($crate::Params::Positional(xs)),
+            // if the object serializes to `Null` map it to `None`
+            $crate::macro_exports::Value::Null => None,
+            // encode any other value as an array of 1 item
+            value => Some($crate::Params::Positional(vec![value])),
```

Comment:
I'm not sure about the semantics here, how this will work out. What if I want to send single positional vec? So, the result should be `[[42]]`.

Maybe if I see some examples, it'll be clearer, but currently I'm afraid about this heuristic being too surprising in some cases.